frame/support/procedural/src/benchmark.rs

```diff
@@ -0,0 +1,836 @@
+						let segment = attr.path.segments.last()?;
+						let _: keywords::block = syn::parse(segment.ident.to_token_stream().into()).ok()?;
+						let mut block = block.clone();
+
+						// consume #[block] tokens
+						block.attrs.remove(k);
+
+						Some(Ok((i, BenchmarkCallDef::Block { block, attr_span: attr.span() })))
+					})
+				},
+				_ => None
+			}
+		}).collect::<Result<Vec<_>>>()?;
+		let (i, call_def) = match &call_defs[..] {
+			[(i, call_def)] => (*i, call_def.clone()), // = 1
+			[] => return Err(Error::new( // = 0
+				item_fn.block.brace_token.span,
+				"No valid #[extrinsic_call] or #[block] annotation could be found in benchmark function body."
+			)),
+			_ => return Err(Error::new( // > 1
+				call_defs[1].1.attr_span(),
+				"Only one #[extrinsic_call] or #[block] attribute is allowed per benchmark."
+			)),
+		};
+
+		Ok(BenchmarkDef {
+			params,
+			setup_stmts: Vec::from(&item_fn.block.stmts[0..i]),
+			call_def,
+			verify_stmts: Vec::from(&item_fn.block.stmts[(i + 1)..item_fn.block.stmts.len()]),
+			extra,
+			skip_meta,
+		})
+	}
+}
+
+/// Parses and expands a `#[benchmarks]` or `#[instance_benchmarks]` invocation
+pub fn benchmarks(attrs: TokenStream, tokens: TokenStream, instance: bool) -> TokenStream {
+	let module = parse_macro_input!(tokens as ItemMod);
+	let mod_span = module.span();
+	let where_clause = match syn::parse::<Nothing>(attrs.clone()) {
+		Ok(_) => quote!(),
+		Err(_) => parse_macro_input!(attrs as WhereClause).predicates.to_token_stream(),
+	};
+	let mod_vis = module.vis;
+	let mod_name = module.ident;
+	let mut expanded_stmts: Vec<TokenStream2> = Vec::new();
+	let mut benchmark_defs: Vec<BenchmarkDef> = Vec::new();
+	let mut benchmark_names: Vec<Ident> = Vec::new();
+	let mut extra_benchmark_names: Vec<Ident> = Vec::new();
+	let mut skip_meta_benchmark_names: Vec<Ident> = Vec::new();
+	let Some((_brace, mut content)) = module.content else {
+		// this will compile error already because attributes attached to empty modules are unstable
+		// but including error anyway to make this future-proof
+		return quote_spanned!(mod_span=> "Module cannot be empty!".to_compile_error()).into()
+	};
+	for stmt in &mut content {
+		let mut push_stmt = || {
+			expanded_stmts.push(stmt.to_token_stream());
+		};
+		let Item::Fn(mut func) = stmt.clone() else { push_stmt(); continue; };
+		for (i, attr) in (&func.attrs.clone()).iter().enumerate() {
+			let Some(seg) = attr.path.segments.last() else { push_stmt(); continue; };
+			let Ok(_) = syn::parse::<keywords::benchmark>(seg.ident.to_token_stream().into()) else { push_stmt(); continue; };
+			let tokens = attr.tokens.to_token_stream().into();
+			let args = parse_macro_input!(tokens as BenchmarkAttrs);
+
+			// consume #[benchmark] attr
+			func.attrs.remove(i);
+
+			// parse benchmark def
+			let benchmark_def = match BenchmarkDef::from(&func, args.extra, args.skip_meta) {
+				Ok(def) => def,
+				Err(err) => return err.to_compile_error().into(),
+			};
```

Comment:
```suggestion
			let benchmark_def = BenchmarkDef::from(&func, args.extra, args.skip_meta).map_err(|e| e.to_compile_error().into())?;
```

Reply:
Actually this one I will have to tweak a bit -- the enclosing function is not a Result